apps/prairielearn/python/zygote.py

```diff
@@ -320,35 +320,19 @@ def worker_loop() -> None:
                 # Any function that is not 'file' or 'render' will modify 'data' and
                 # should not be returning anything (because 'data' is mutable).
                 if (fcn != "file") and (fcn != "render"):
-                    if val is None:
+                    if val is None or val is args[-1]:
                         json_outp = try_dumps(
                             {"present": True, "val": args[-1]}, allow_nan=False
                         )
                     else:
-                        json_outp_passed = try_dumps(
-                            {"present": True, "val": args[-1]},
-                            sort_keys=True,
-                            allow_nan=False,
-                        )
-                        json_outp = try_dumps(
-                            {"present": True, "val": val},
-                            sort_keys=True,
-                            allow_nan=False,
+                        # We'll only actually complain if the function returned
+                        # a completely different object than the one passed in.
+                        # Otherwise, we'll just silently ignore the return value
+                        # and use the passed-in object (which should in fact be
+                        # the same object).
+                        raise Exception(
+                            f"Function {str(fcn)} in file {str(file)} returned a data object other than the one that was passed in"
```

Comment:
I do really want to rip the bandaid off here, but: maybe we should instead continue warning (not erroring) a little longer, but this time with the revised check of referential equality? It's certainly possible that someone is doing something weird like `return copy.deepcopy(data)`, which the old check wouldn't pick up but this one would.

Reply:
At some level, it doesn't matter _that_ much whether an exception is raised or a warning is emitted, they will still both result in the same kind of issue being raised to the instructor (though the exception will make it so the student can't complete the given question). It's probably ok to leave this as an exception.

I've mentioned this a few times, but this is another good use case for deprecation-style warnings that are only shown to the instructor: https://github.com/PrairieLearn/PrairieLearn/issues/7337

Reply:
Yeah, the "student can't complete the question" is the bit that I'm most worried about, and why I'm wondering if we should err on the side of caution and avoid raising an exception for at least a week or two.

Reply:
If we want to be extra careful, I think raising a warning is the way to go. The end of the semester is rolling around soon, and it would make sense to me to deploy a hard-error after finals. Ultimately depends on how much time an extra PR + deployment for this takes up.

Reply:
True, I forgot about finals; it's probably best to switch this to a warning for now. We're _definitely_ not going to land `orjson` before finals.

Reply:
I switched back to a warning in 6c0b65b.